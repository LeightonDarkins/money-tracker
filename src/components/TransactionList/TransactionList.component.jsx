import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Transaction from '../Transaction/Transaction.component.jsx'

class TransactionList extends Component {
  constructor (props) {
    super(props)

    this.routeMatch = this.props.match
  }

  componentDidMount () {
    this.props.fetchCategories()

    if (this.routeMatch && this.routeMatch.params) {
      this.props.fetchTransactions(this.routeMatch.params.accountId)
    }
  }

  bodyContent () {
    if (this.props.isLoading) {
      return (<div className='transaction-list-spinner-container'>
        <i className='fa fa-circle-o-notch fa-spin fa-2x transaction-list-spinner' aria-hidden='true' />
      </div>)
    }

    if (this.props.transactions.length === 0) return <ul><li>No Transactions For Account</li></ul>

    return <ul id='TransactionList'>
      {
        this.props.transactions.map(transaction => {
          this.props.categories.forEach(category => {
            if (transaction.category === category.id) {
              transaction.category = category.name
            }
          })

          return <Transaction
            key={transaction._id}
            amount={transaction.amount}
            date={transaction.date}
            category={transaction.category}
            onClick={() => this.props.onTransactionClick(transaction.id)} />
        })
      }
    </ul>
  }

  render () {
    return (
      <div>
        <section className='mt-header'>
          <h2>Transactions</h2>
        </section>
        <section className='mt-body'>
          { this.bodyContent() }
        </section>
        <section className='mt-footer'>
          footer
        </section>
      </div>
    )
  }
}

TransactionList.propTypes = {
  transactions: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string.isRequired,
      amount: PropTypes.number.isRequired,
      date: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired
    })
  ).isRequired,
  onTransactionClick: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  fetchTransactions: PropTypes.func.isRequired,
  fetchCategories: PropTypes.func.isRequired
}

export default TransactionList
