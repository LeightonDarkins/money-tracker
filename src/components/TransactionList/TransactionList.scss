@import '../../variables.scss';

#TransactionList {
  .scroll-view {
    overflow: auto;
    list-style-type: none;
    padding: 0;
    margin: 0;
    height: 100%;
  }

  .spinner-container, .placeholder {
    text-align: center;
    margin-top: $placeholder-spacing;
  }

  .date-divider {
    padding: $default-spacing;
    margin: 0 auto 0 auto;
    text-align: center;
    background-color: #efefef;
    border-bottom: 1px solid #ddd;
  }
}
