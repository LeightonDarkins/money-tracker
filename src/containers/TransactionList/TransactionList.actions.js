export const types = {
  fetchTransactionsStarted: 'FETCH_TRANSACTIONS_STARTED',
  fetchTransactionsSucceeded: 'FETCH_TRANSACTIONS_SUCCEEDED',
  fetchTransactionsFailed: 'FETCH_TRANSACTIONS_FAILED',
  transactionClicked: 'TRANSACTION_CLICKED',
  fetchTransactions: 'FETCH_TRANSACTIONS',
  fetchCategories: 'TRANSACTION_LIST_FETCH_CATEGORIES',
  fetchCategoriesSucceeded: 'TRANSACTION_LIST_FETCH_CATEGORIES_SUCCEEDED'
}

export const fetchTransactionsStarted = () => ({ type: types.fetchTransactionsStarted })

export const fetchTransactionsFailed = () => ({ type: types.fetchTransactionsFailed })

export const fetchTransactionsSucceeded = (transactions) => ({ type: types.fetchTransactionsSucceeded, transactions })

export const transactionClicked = (id) => ({ type: types.transactionClicked, id })

export const fetchTransactions = (id) => ({ type: types.fetchTransactions, accountId: id })

export const fetchCategories = () => ({ type: types.fetchCategories })

export const fetchCategoriesSucceeded = (categories) => ({ type: types.fetchCategoriesSucceeded, categories })
