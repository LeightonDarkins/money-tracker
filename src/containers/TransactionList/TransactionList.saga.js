import { put, call, takeLatest } from 'redux-saga/effects'
import AccountApi from '../../apis/Account.api'
import CategoryApi from '../../apis/Category.api'
import { fetchTransactionsSucceeded,
  fetchTransactionsStarted,
  fetchTransactionsFailed,
  fetchCategoriesSucceeded,
  types } from './TransactionList.actions'
import { apiError } from '../../common/common.actions'

export function * fetchTransactions (action) {
  try {
    yield put(fetchTransactionsStarted())
    const transactions = yield call(AccountApi.fetchTransactionsForAccount, action.accountId)
    yield put(fetchTransactionsSucceeded(transactions))
  } catch (error) {
    yield put(apiError(error))
    yield put(fetchTransactionsFailed())
  }
}

export function * fetchCategories () {
  try {
    const categories = yield call(CategoryApi.fetchCategories)
    yield put(fetchCategoriesSucceeded(categories))
  } catch (error) {
    yield put(apiError(error))
  }
}

function * accountListSaga () {
  yield takeLatest(types.fetchTransactions, fetchTransactions)
  yield takeLatest(types.fetchCategories, fetchCategories)
}

export default accountListSaga
