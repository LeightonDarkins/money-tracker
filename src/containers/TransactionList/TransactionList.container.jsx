import { connect } from 'react-redux'
import TransactionList from '../../components/TransactionList/TransactionList.component.jsx'
import { fetchTransactions, fetchCategories } from './TransactionList.actions'

const mapStateToProps = state => {
  return {
    transactions: state.TransactionList.transactions,
    categories: state.TransactionList.categories,
    isLoading: state.TransactionList.isLoading
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onTransactionClick: (id) => {
      console.log(`Transaction Clicked ${id}`)
    },
    fetchTransactions: (id) => {
      dispatch(fetchTransactions(id))
    },
    fetchCategories: (id) => {
      dispatch(fetchCategories())
    }
  }
}

const ConnectedTransactionList = connect(
  mapStateToProps,
  mapDispatchToProps)(TransactionList)

export default ConnectedTransactionList
