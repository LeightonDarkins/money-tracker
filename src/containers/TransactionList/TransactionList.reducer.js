import { types } from './TransactionList.actions'

const initialState = { transactions: [], categories: [], isLoading: false }

const accounts = (state = initialState, action) => {
  switch (action.type) {
    case types.fetchTransactionsStarted:
      return Object.assign({}, state, { isLoading: true })
    case types.fetchTransactionsSucceeded:
      return Object.assign({}, state,
        {
          transactions: action.transactions,
          isLoading: false
        })
    case types.fetchTransactionsFailed:
      return Object.assign({}, state, { isLoading: false })
    case types.fetchCategoriesSucceeded:
      return Object.assign({}, state, { categories: action.categories })
    default:
      return state
  }
}

export default accounts
