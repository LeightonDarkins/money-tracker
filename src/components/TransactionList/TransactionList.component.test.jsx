/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import TransactionList from './TransactionList.component'

describe('TransactionListComponent', () => {
  const transactions = [
    {
      id: 'transaction-1',
      amount: 1000,
      date: '2017-11-08T20:42:35.805Z',
      category: 'category-1'
    }
  ]

  describe('Snapshots', () => {
    it('renders correctly with no transactions', () => {
      const transactionList = renderer.create(<TransactionList
        transactions={[]}
        onTransactionClick={(y) => y}
        />)

      expect(transactionList).toMatchSnapshot()
    })

    it('renders correctly while loading', () => {
      const transactionList = renderer.create(<TransactionList
        transactions={[]}
        onTransactionClick={(y) => y}
        isLoading
        />)

      expect(transactionList).toMatchSnapshot()
    })

    it('renders correctly with transactions', () => {
      const transactionList = renderer.create(<TransactionList
        transactions={transactions}
        onTransactionClick={(y) => y}
        />)

      expect(transactionList).toMatchSnapshot()
    })
  })

  describe('Behaviour', () => {
    let mockAddTransactionClick

    beforeEach(() => {
      mockAddTransactionClick = jest.fn()
    })

    afterEach(() => {
      mockAddTransactionClick.mockReset()
    })

    it('calls onTransactionClick when a transaction is clicked', () => {
      const wrapper = shallow(<TransactionList
        transactions={transactions}
        onTransactionClick={mockAddTransactionClick}
        isLoading={false} />)

      let transactionItems = wrapper.find('#TransactionList Transaction')

      expect(transactionItems).toHaveLength(1)

      transactionItems.simulate('click')

      expect(mockAddTransactionClick).toBeCalled()
    })
  })
})
