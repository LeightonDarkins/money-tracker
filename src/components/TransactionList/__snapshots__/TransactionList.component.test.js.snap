exports[`TransactionListComponent Snapshots renders correctly while loading 1`] = `
<div>
  <section
    className="mt-header">
    <h2>
      Transactions
    </h2>
  </section>
  <section
    className="mt-body">
    <div
      className="transaction-list-spinner-container">
      <i
        aria-hidden="true"
        className="fa fa-circle-o-notch fa-spin fa-2x transaction-list-spinner" />
    </div>
  </section>
  <section
    className="mt-footer">
    footer
  </section>
</div>
`;

exports[`TransactionListComponent Snapshots renders correctly with no transactions 1`] = `
<div>
  <section
    className="mt-header">
    <h2>
      Transactions
    </h2>
  </section>
  <section
    className="mt-body">
    <ul>
      <li>
        No Transactions For Account
      </li>
    </ul>
  </section>
  <section
    className="mt-footer">
    footer
  </section>
</div>
`;

exports[`TransactionListComponent Snapshots renders correctly with transactions and categoriees 1`] = `
<div>
  <section
    className="mt-header">
    <h2>
      Transactions
    </h2>
  </section>
  <section
    className="mt-body">
    <ul
      id="TransactionList">
      <li
        onClick={[Function]}>
        First Category
         : 
        1000
         : 
        2017-11-08T20:42:35.805Z
      </li>
    </ul>
  </section>
  <section
    className="mt-footer">
    footer
  </section>
</div>
`;

exports[`TransactionListComponent Snapshots renders correctly with transactions, no categories 1`] = `
<div>
  <section
    className="mt-header">
    <h2>
      Transactions
    </h2>
  </section>
  <section
    className="mt-body">
    <ul
      id="TransactionList">
      <li
        onClick={[Function]}>
        category-1
         : 
        1000
         : 
        2017-11-08T20:42:35.805Z
      </li>
    </ul>
  </section>
  <section
    className="mt-footer">
    footer
  </section>
</div>
`;
